### DISCUSSION TITLE
REST API: Query error: worker call timed out.

### DISCUSSION THREAD

**Discussion by Leandrodn** [Budibase/budibase#3847] [Help]
Hi, I am new to budibase and would like to use it for my internal admin dashboard. I have a REST API made in Python Flask. 
When is tried to test the API connection, I get a pop-up with query error: worker call timed out. 
![image](https://user-images.githubusercontent.com/81325604/146936703-bbe76e6a-a596-4cac-845e-660337032d1a.png)
The API is running and returns JSON data. When I test the API in Postman, everything is fine. 
![image](https://user-images.githubusercontent.com/81325604/146937044-4300eac9-82cc-49d2-8c30-54a25a05789c.png)
The URL is the same as in Postman. 
![image](https://user-images.githubusercontent.com/81325604/146937110-d384f0d8-5f2b-4e77-be2c-c6f272e62af0.png)
![image](https://user-images.githubusercontent.com/81325604/146937215-51a04d63-d8fa-4a25-9c96-86af06cf4446.png)

**Comment by mjashanks** [comment]
Hi @Leandrodn 

How are you running Budibase? 
- Are you using our cloud platform: `https://your-domain.budibase.app`?
- Are you self-hosting via Docker? Where is Budibase running? I.e. on your machine (`localhost`) or in a cloud platform somewhere?

A timeout would suggest that might need to whitelist the IP that Budibase is running on - but that should not be the issue if running on the same machine as postman.

Budibase Cloud IP address for whitelisting: https://github.com/Budibase/budibase/discussions/3114

**Reply by Leandrodn** [reply]
HI @mjashanks,
I am using the budibase cloud platform for now. The API is running on a remote server. I whitelisted the budibase IP, but the error still occurred. Is it possible that this error occurs because I'm using a self generated SSL-certificate.

**Comment by shogunpurple** [comment]
Hi @Leandrodn - I just tested this and received the same result. The HTTP calls are made from the backend of the budibase application, so it should be a similar flow to just using postman. 

I just tried hitting your endpoint from hoppscotch as well (https://hoppscotch.io/) and it timed out there too - so I don't think this is budibase specific. I think there must be a networking issue on your side, where requests from outside the network are being blocked by your firewall. Did you open up port 443 to the budibase IP?

Are you using a cloud provider or is this just running inside your network? An iplookup shows your API is running somewhere in Rotterdam - how are you allowing IP addresses to access your server? `ufw` or something similar?

**Comment by Gijs-Kortlever** [comment]
Hi @shogunpurple,
I work on this project with @Leandrodn.
We are using ufw to allow IP-addresses access to our ubuntu server, there we are using NGINX as reverse proxy server with a self-generated ssl certificate, all HTTP calls are redirected to the HTTPS side of the NGINX server. 
We added the IP-address mentioned by @mjashanks to the ufw rules, so budibase should be allowed to make calls to NGINX and indirectly the API. 
Port 443 on our server should be completely open for HTTPS requests, so i am struggling to find the cause of this.
I looked at (https://hoppscotch.io) as well and it worked fine, so from our side it seems to all be working.

Hoppscotch:
![afbeelding](https://user-images.githubusercontent.com/72206245/147150179-3ff776e0-0cea-4067-931c-ca597d64e063.png)

Our server is running remotely we are not on the same network as the server itself, so as it seems to be working for us in theory it should also work for you and budibase.

ufw rules:
![afbeelding](https://user-images.githubusercontent.com/72206245/147150653-29d1af1c-fa7f-43a0-aa63-c61a22cf1f80.png)

I disabled HTTPS to test if HTTP might work but again very conflicting responses. Postman is working fine, we can even use the API URL in Excel but it refuses to work in budibase.
It also doesn't work on Hoppscotch but that is because Hoppscotch doesn't accept HTTP URL's.

Postman:
![afbeelding](https://user-images.githubusercontent.com/72206245/147153567-a6221438-2a44-456b-9832-d15b752d5b6b.png)

**Reply by shogunpurple** [reply]
Hi @Gijs-Kortlever,

It did not work in hoppscotch for me - it keeps timing out. Is there a header or some other setting that you have there that I don't have? Since hoppscotch is a hosted service, I would think you would see the same result as me.

![Screenshot 2021-12-23 at 10 12 14](https://user-images.githubusercontent.com/11256663/147225329-e777e604-6e3c-4801-8f9d-d9c1c31b8114.png)


`curl` also timed out - I tried running from my machine, and from EC2 instances running in `eu-west-1`, including the instances our production cluster is running on, which you have whitelisted. All of them timed out:
`curl https://145.24.222.39/register/retrieve`
`curl http://145.24.222.39/register/retrieve`

Hoppscotch works with HTTP Urls - here it is hitting a webpage over HTTP. 
![Screenshot 2021-12-23 at 09 55 48](https://user-images.githubusercontent.com/11256663/147223206-203926ea-f10d-4403-a749-3fdf206a719c.png)

But using this with `http://145.24.222.39/register/retrieve` still times out.


Postman is also timing out for me:

![Screenshot 2021-12-23 at 10 11 21](https://user-images.githubusercontent.com/11256663/147225220-d5c1780e-c77a-48ea-8536-3cffbefdfc93.png)

If I was able to hit this API with `curl` or postman from anywhere, I would say that the problem would be with budibase, but it really seems like this API is inaccessible. There's nothing that seems to be happening in the code that would cause this either - self signed certs are supported and shouldn't interfere.

Is there a chance that there is some sort of firewall rule that is blocking us? Location based, or similar?

**Reply by shogunpurple** [reply]
@Gijs-Kortlever does it work if you try a different network? For example using mobile data and hitting that URL on your phone or something. It did not work for me when I did that either.

**Comment by ssuryansh** [comment]
@mjashanks Hi I have deployed budiabse on my aws ec2 instance by using docker container and i want to access below api from budibase  and this is working via postman but not working from rest api component of budiabse.
My aws ec2 instance public ip is- 54.165.164.159.

curl --location --request POST 'http://54.165.164.159:10000/api/ta_0d44360bb7e341c8be830988b62fea12/search/' \
--header 'User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:89.0) Gecko/20100101 Firefox/89.0' \
--header 'Accept: application/json' \
--header 'Accept-Language: en-GB,en;q=0.5' \
--header 'Referer: http://localhost:10000/app_dev_887f393b1c4b4084b888afa7d0d109cf' \
--header 'x-budibase-app-id: app_dev_b4184a827c8d4a9b878006a80e03d4d7' \
--header 'Content-Type: application/json' \
--header 'x-budibase-type: client' \
--header 'Origin: http://localhost:10000/' \
--header 'Connection: keep-alive' \
--header 'x-budibase-api-key: fe833bf671b7d4c63ebd03450d463de7-f835e26b8f4ca5512805aa8158123dc8bc026ca5f379a178b5c7b8dd02b27c84b421508ed78e166f' \
--header 'Cookie: budibase:currentapp=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJhcHBJZCI6ImFwcF9kZXZfY2Q0MzUzYzE5ZjFmNDUxYWFhZTI3ZGNmYzA4MDU1ODciLCJpYXQiOjE2NTA1MzcxNjN9.rqx8dEKkrvwUikbLf8FkbYoZqB_TqLD9GoH23Hf47z8; budibase:currentapp=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJhcHBJZCI6ImFwcF9kZXZfYjQxODRhODI3YzhkNGE5Yjg3ODAwNmE4MGUwM2Q0ZDciLCJpYXQiOjE2NTI5NjExMzV9.AMLNBdOMWRK-6WUnpOpzE9Q8Ebg9vRwoXrOXisjEZKQ' \
--data-raw '{
    "query":{
        "equal": {
            "package_name": "test",
            "packageAction": "Install"
        }
    }
}'

Erroe from budibase rest api component -
**Error: worker call timed out!**

Is there any config updation needed on aws console or ec2 server or any othe place.Please let me know?
